### PR TITLE
Fixed pruning buffer is full error in import db

### PR DESCRIPTION
**Pull request by iulianpascalau**
- fixed `pruning buffer is full` error
- improved user account checking

### CODE REVIEW

**Review by BeniaminDrasovean — DISMISSED** [review]

**Review by bogdan-rosianu — APPROVED** [review]
:+1: :partying_face:

File: process/smartContract/hooks/blockChainHook.go [hunk 1]
```diff
@@ -397,16 +401,18 @@ func (bh *BlockChainHookImpl) IsPayable(address []byte) (bool, error) {
 		return true, nil
 	}
 
+	if bh.isNotSystemAccountAndCrossShard(address) {
+		//backwards compatibility
```
Comment (sasurobert) [inline process/smartContract/hooks/blockChainHook.go:405]: no need for comment

Comment (iulianpascalau) [reply process/smartContract/hooks/blockChainHook.go:405]: removed

**Review by sasurobert — DISMISSED** [review]

**Review by bogdan-rosianu — DISMISSED** [review]

File: state/storagePruningManager/evictionWaitingList/memoryEvictionWaitingList.go [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+	defer mewl.opMutex.RUnlock()
+
+	info, found := mewl.reversedCache[hash]
+	if !found {
+		return false, nil
+	}
+
+	for _, key := range info.roothashes {
+		if len(key) == 0 {
+			return false, state.ErrInvalidKey
```
Comment (bogdan-rosianu) [inline state/storagePruningManager/evictionWaitingList/memoryEvictionWaitingList.go:150]: wrap error and print the key?

File: state/storagePruningManager/evictionWaitingList/memoryEvictionWaitingList.go [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+
+	mewl.putInReversedCache(rootHash, hashes)
+
+	if uint(len(mewl.cache)) <= mewl.cacheSize {
+		mewl.cache[string(rootHash)] = hashes
+		return nil
+	}
+
+	log.Warn("trie nodes eviction waiting list full, emptying...")
+	mewl.cache = make(map[string]temporary.ModifiedHashes)
```
Comment (bogdan-rosianu) [inline state/storagePruningManager/evictionWaitingList/memoryEvictionWaitingList.go:61]: why do we clear the entire cache on Put?

**Review by bogdan-rosianu — CHANGES_REQUESTED** [review]

File: state/storagePruningManager/evictionWaitingList/memoryEvictionWaitingList.go [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+// Put stores the given hashes in the eviction waiting list, in the position given by the root hash
+func (mewl *memoryEvictionWaitingList) Put(rootHash []byte, hashes temporary.ModifiedHashes) error {
+	mewl.opMutex.Lock()
+	defer mewl.opMutex.Unlock()
+
+	log.Trace("trie eviction waiting list", "size", len(mewl.cache))
+
+	mewl.putInReversedCache(rootHash, hashes)
+
+	if uint(len(mewl.cache)) <= mewl.cacheSize {
```
Comment (sasurobert) [inline state/storagePruningManager/evictionWaitingList/memoryEvictionWaitingList.go:55]: can we have a similar check for reversedCache - just to make sure that no leak happens ? even if someone later changes the codes.

File: state/storagePruningManager/evictionWaitingList/memoryEvictionWaitingList.go (hunk 3 above)
Comment (iulianpascalau) [reply state/storagePruningManager/evictionWaitingList/memoryEvictionWaitingList.go:61]: As agreed, when the pruning buffer is full, we should empty it in order to be able to process the future pruning hashes.

File: state/storagePruningManager/evictionWaitingList/memoryEvictionWaitingList.go (hunk 2 above)
Comment (iulianpascalau) [reply state/storagePruningManager/evictionWaitingList/memoryEvictionWaitingList.go:150]: the key is empty. What should I print?

File: factory/stateComponents.go [hunk 5]
```diff
@@ -161,27 +157,9 @@ func (scf *stateComponentsFactory) createPeerAdapter() (state.AccountsAdapter, e
...
-	}
-
-	evictionDb, err := storageUnit.NewDB(arg)
-	if err != nil {
-		return nil, err
-	}
-
-	trieEvictionWaitingList, err := evictionWaitingList.NewEvictionWaitingList(
+func (scf *stateComponentsFactory) newStoragePruningManager() (state.StoragePruningManager, error) {
+	trieEvictionWaitingList, err := evictionWaitingList.NewMemoryEvictionWaitingListV2(
```
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply factory/stateComponents.go:161]: Done, thanks

File: state/storagePruningManager/evictionWaitingList/memoryEvictionWaitingList.go (hunk 4 above)
Comment (iulianpascalau) [reply state/storagePruningManager/evictionWaitingList/memoryEvictionWaitingList.go]: done, added another config value to be more extensible

**Review by sasurobert — APPROVED** [review]

**Review by bogdan-rosianu — APPROVED** [review]

**Pull request merged**